### PR TITLE
Fix: Refactor round logic, fix round starting bug, fix quickshop exception

### PR DESCRIPTION
**Pull request by mzegar**
- Going to run this through some testing
- I managed to repro a similar bug to #321 so this might fix it, if not, it fixes a different bug
- Fixes an exception in quickshop

### CODE REVIEW

File: code/rounds/InProgressRound.cs [hunk 1]
```diff
@@ -60,9 +59,9 @@ protected override void OnStart()
             {
                 foreach (Client client in Client.All)
                 {
-                    if (client.Pawn is not TTTPlayer player)
+                    if (client is not TTTPlayer player)
                     {
-                        continue;
+                        return;
                     }
```
Comment (Alf21) [inline code/rounds/InProgressRound.cs:65]: This is always true

File: code/rounds/InProgressRound.cs [hunk 2]
```diff
@@ -60,9 +59,9 @@ protected override void OnStart()
             {
                 foreach (Client client in Client.All)
                 {
-                    if (client.Pawn is not TTTPlayer player)
+                    if (client is not TTTPlayer player)
                     {
-                        continue;
+                        return;
```
Comment (Alf21) [inline code/rounds/InProgressRound.cs:64]: sure you wanna return?

File: code/rounds/InProgressRound.cs [hunk 3]
```diff
@@ -175,11 +161,11 @@ private TTTTeam IsRoundOver()
         private void AssignRoles()
         {
             // TODO: There should be a neater way to handle this logic.
-            int traitorCount = (int) Math.Max(Players.Count * 0.25f, 1f);
+            int traitorCount = (int) Math.Max(_startingRoundPlayers.Count * 0.25f, 1f);
 
             for (int i = 0; i < traitorCount; i++)
             {
-                List<TTTPlayer> unassignedPlayers = Players.Where(p => p.Role is NoneRole).ToList();
+                List<TTTPlayer> unassignedPlayers = _startingRoundPlayers.Where(p => p.Role is NoneRole).ToList();
```
Comment (Alf21) [inline code/rounds/InProgressRound.cs:168]: make sure, these players still exists (could have disconnected meanwhile)

Comment (mzegar) [reply code/rounds/InProgressRound.cs]: Let me know if what I have now is fine.

File: code/rounds/InProgressRound.cs [hunk 4]
```diff
@@ -201,6 +187,8 @@ private void AssignRoles()
                     player.SendClientRole();
                 }
             }
+
+            _areRolesAssigned = true;
```
[the remark this answers is not in the document]
Comment (mzegar) [reply code/rounds/InProgressRound.cs:174]: Yeah, so the death of a player can trigger this event, I think that's why #321 occured.

File: code/ui/alivehud/damageindicator/DamageIndicator.cs [hunk 5]
```diff
@@ -22,10 +22,9 @@ public DamageIndicator() : base()
             Instance = this;
 
             StyleSheet.Load("/ui/alivehud/damageindicator/DamageIndicator.scss");
-
             Style.SetBackgroundImage(Texture.Load("/ui/damageindicator/default.png"));
-
             Style.ZIndex = -1;
+            Style.Dirty();
```
Comment (mzegar) [inline code/ui/alivehud/damageindicator/DamageIndicator.cs:27]: I think a s&box bug to log here, I'm getting different behavior with this here.

File: code/rounds/InProgressRound.cs [hunk 6]
```diff
@@ -75,11 +75,7 @@ protected override void OnStart()
                         player.SetHealth(player.MaxHealth);
                     }
 
-                    if (!player.IsForcedSpectator)
-                    {
-                        SetLoadout(player);
-                        _startingRoundPlayers.Add(player);
-                    }
+                    SetLoadout(player);
```
Comment (Alf21) [inline code/rounds/InProgressRound.cs:78]: I'd still give the loadout on spawn regarding death and the possibility searching other weapons which making it more obvious and leading to more weapons in game

File: code/rounds/InProgressRound.cs [hunk 7]
```diff
@@ -160,12 +142,13 @@ private TTTTeam IsRoundOver()
...
-            // TODO: There should be a neater way to handle this logic.
-            int traitorCount = (int) Math.Max(_startingRoundPlayers.Count * 0.25f, 1f);
+            List<TTTPlayer> alivePlayers = Utils.GetAlivePlayers();
+
+            int traitorCount = (int) Math.Max(alivePlayers.Count * 0.25f, 1f);
 
             for (int i = 0; i < traitorCount; i++)
             {
-                List<TTTPlayer> unassignedPlayers = _startingRoundPlayers.Where(p => p.Role is NoneRole).ToList();
+                List<TTTPlayer> unassignedPlayers = alivePlayers.Where(p => p?.Role is NoneRole).ToList();
```
Comment (Alf21) [inline code/rounds/InProgressRound.cs:151]: Did you tested this? Are forced spectators alive? And why `p?`

File: code/ui/generalhud/inspectmenu/InspectMenu.cs [hunk 8]
```diff
@@ -211,7 +211,13 @@ private void UpdateCurrentInspectDescription()
...
+            if (!_playerCorpse.IsValid())
+            {
+                Enabled = false;
+                return;
+            }
+
+            if (Enabled && _playerCorpse.Transform.Position.Distance(Local.Pawn.Owner.Position) > 100f)
             {
                 Enabled = false;
             }
```
Comment (Alf21) [inline code/ui/generalhud/inspectmenu/InspectMenu.cs:223]: can't this be combined?

File: code/rounds/InProgressRound.cs (hunk 7 above)
Comment (mzegar) [reply code/rounds/InProgressRound.cs:151]: Removed `?`, yeah I've been doing some intensive testing making sure this large logic change doesn't mess things up.

File: code/rounds/InProgressRound.cs (hunk 6 above)
Comment (mzegar) [reply code/rounds/InProgressRound.cs:78]: Hmm not sure if I fully understand, could you rephrase or PM me?

Previously we were setting the loadout twice which I want to avoid. The `SetLoadout(...)` call also happens right after we either respawn the player or set their hp to max.

File: code/rounds/InProgressRound.cs [hunk 9]
```diff
@@ -156,7 +124,7 @@ private TTTTeam IsRoundOver()
         {
             List<TTTTeam> aliveTeams = new();
 
-            foreach (TTTPlayer player in Players)
+            foreach (TTTPlayer player in Utils.GetAlivePlayers())
```
Comment (Alf21) [inline code/rounds/InProgressRound.cs:127]: TTT also has add-ons like spectator deathmatch, which is a mod that adds the ability to offer dead players the possibility to join a deathmatch while others are still fighting.

With your change, this won't be possible anymore as the specdm player would be alive as well. That's additionally why I'd like to get track of the current player in a round

File: code/rounds/InProgressRound.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (Alf21) [reply code/rounds/InProgressRound.cs:174]: Wouldn't it be better then to assign the roles in PreparingRound before switching to the inprogress round (or setting it as the current one), to keep it clean

File: code/ui/generalhud/inspectmenu/InspectMenu.cs [hunk 10]
```diff
@@ -211,9 +211,10 @@ private void UpdateCurrentInspectDescription()
 
         public override void Tick()
         {
-            if (Enabled && _playerCorpse?.Transform.Position.Distance(Local.Pawn.Owner.Position) > 100f)
+            if (!Enabled || !_playerCorpse.IsValid() || _playerCorpse.Transform.Position.Distance(Local.Pawn.Owner.Position) > 100f)
             {
                 Enabled = false;
+                return;
```
Comment (Alf21) [inline code/ui/generalhud/inspectmenu/InspectMenu.cs:217]:
```suggestion

                return;
```

File: code/rounds/PreRound.cs [hunk 11]
```diff
@@ -63,7 +129,14 @@ private static async Task StartRespawnTimer(TTTPlayer player)
...
         {
-            AddPlayer(player);
+            if (player.IsForcedSpectator)
+            {
+                return;
+            }
+
+            Extensions.Log.Debug($"Added Hands to {player.Client.Name}");
+
+            player.Inventory.TryAdd(new Hands(), true);
```
Comment (mzegar) [inline code/rounds/PreRound.cs:139]: @Alf21 How do we want to handle this, we end up calling `Player.Respawn()` on everyone (including forced spectators) which gives the spectators `Hands()`.

1. Inside of `Respawn()` we handle the `IsForcedSpectator` case, so I think we still want to call `Respawn()` on them
2. We probably want to be able to give items to dead players, so no internal check inside of `TryAdd()`

File: code/map/MapHandler.cs [hunk 12]
```diff
@@ -29,6 +31,10 @@ public MapHandler()
                 {
                     ModelEntityDataList.Add(ModelEntityData.Create(entity as ModelEntity));
                 }
+                else if (entity is TTTWeaponRandom)
+                {
+                    RandomWeaponCount += 1;
```
Comment (Alf21) [inline code/map/MapHandler.cs:36]:
```suggestion
                    RandomWeaponCount++;
```

File: code/rounds/InProgressRound.cs [hunk 13]
```diff
@@ -15,10 +15,14 @@
...
 namespace TTTReborn.Rounds
 {
-    public class InProgressRound : BaseRound
+    public partial class InProgressRound : BaseRound
     {
         public override string RoundName => "In Progress";
-        private List<TTTLogicButton> LogicButtons;
+
+        [Net] public List<TTTPlayer> Players { get; set; }
+        [Net] public List<TTTPlayer> Spectators { get; set; }
```
Comment (Alf21) [inline code/rounds/InProgressRound.cs:23]:
```suggestion
        [Net]
        public List<TTTPlayer> Players { get; set; }

        [Net]
        public List<TTTPlayer> Spectators { get; set; }
```

File: code/rounds/PreRound.cs (hunk 11 above)
Comment (Alf21) [reply code/rounds/PreRound.cs:139]: We should just add a check in the TryAdd method to never give a weapon to a non alive player.

File: code/rounds/InProgressRound.cs [hunk 14]
```diff
@@ -152,13 +100,20 @@ private static void SetLoadout(TTTPlayer player)
...
+        }
+
         private TTTTeam IsRoundOver()
         {
             List<TTTTeam> aliveTeams = new();
 
             foreach (TTTPlayer player in Players)
             {
-                if (player.Team != null && !aliveTeams.Contains(player.Team))
+                if (player.LifeState == LifeState.Alive && player.Team != null && !aliveTeams.Contains(player.Team))
```
Comment (Alf21) [inline code/rounds/InProgressRound.cs:116]: Seems like disconnected players aren't handled?

File: code/player/Inventory.cs [hunk 15]
```diff
@@ -87,7 +87,7 @@ public bool Add(IItem item, bool makeActive = false)
         /// <returns></returns>
         public bool TryAdd(IItem item, bool makeActive = false)
         {
-            if (!Add(item, makeActive))
+            if (Owner.LifeState == LifeState.Dead || !Add(item, makeActive))
```
Comment (mzegar) [inline code/player/Inventory.cs:90]: @Alf21 Maybe we want to check for != LifeState.Alive here? 

There are `Dying, Respawnable, Respawning`, I wonder if we should avoid trying to give during these as well.

Comment (Alf21) [reply code/player/Inventory.cs:90]: yea, that should be the way to go

File: code/rounds/PreRound.cs (hunk 11 above)
Comment (mzegar) [reply code/rounds/PreRound.cs]: Done, mentioned you in another comment.

**Review by Alf21 — APPROVED** [review]
